File: Sources/CodeRim/Sources/CodexJSONLParser.swift
import CoreFoundation
import Foundation

enum CodexParsedLine: Equatable, Sendable {
    case sessionMetadata(SessionMetadata)
    case taskStarted(TaskStartedMetadata)
    case turnContext(TurnContextMetadata)
    case imageAttachments(ImageAttachmentMetadata)
    case token(CodexTokenObservation)
    case ignored
    case malformed(String)
}

struct CodexTokenObservation: Equatable, Sendable {
    let occurredAt: Date
    let ordinal: Int64?
    let lastUsage: TokenUsage?
    let cumulativeUsage: TokenUsage?
}

struct TurnContextMetadata: Equatable, Sendable {
    let model: String?
    let workingDirectory: String?
}

struct TaskStartedMetadata: Equatable, Sendable {
    let occurredAt: Date
    let startedAt: Date?
    let ordinal: Int64?
}

struct ImageAttachmentMetadata: Equatable, Sendable {
    let occurredAt: Date
    let count: Int
}

struct CodexJSONLParser: Sendable {
    static let maximumLineBytes = 1_048_576
    static let maximumTokenComponent: Int64 = 1_000_000_000_000
    private static let maximumIdentifierLength = 256
    private static let maximumModelLength = 256
    private static let maximumPathLength = 4_096

    func parse(_ line: Data) -> CodexParsedLine {
        guard !line.isEmpty else { return .ignored }
        guard line.count <= Self.maximumLineBytes else {
            return .malformed("line exceeds the 1 MiB safety limit")
        }

        let object: Any
        do {
            object = try JSONSerialization.jsonObject(with: line)
        } catch {
            return .malformed("invalid JSON")
        }

        guard let root = object as? [String: Any], let type = root["type"] as? String else {
            return .malformed("missing event type")
        }

        switch type {
        case "session_meta":
            return parseSessionMetadata(root)
        case "turn_context":
            return parseTurnContext(root)
        case "event_msg":
            return parseEventMessage(root)
        case "response_item":
            return parseResponseItem(root)
        default:
            return .ignored
        }
    }

    private func parseResponseItem(_ root: [String: Any]) -> CodexParsedLine {
        guard let payload = root["payload"] as? [String: Any],
              payload["type"] as? String == "message",
              payload["role"] as? String == "user",
              let content = payload["content"] as? [[String: Any]]
        else { return .ignored }

        let imageCount = content.reduce(into: 0) { count, item in
            if item["type"] as? String == "input_image" {
                count += 1
            }
        }
        guard imageCount > 0 else { return .ignored }
        guard let timestamp = root["timestamp"] as? String,
              let occurredAt = parseTimestamp(timestamp) else {
            return .malformed("image metadata is missing a valid timestamp")
        }
        return .imageAttachments(ImageAttachmentMetadata(occurredAt: occurredAt, count: imageCount))
    }

    private func parseTurnContext(_ root: [String: Any]) -> CodexParsedLine {
        guard let payload = root["payload"] as? [String: Any] else {
            return .malformed("turn context is missing payload")
        }
        return .turnContext(
            TurnContextMetadata(
                model: sanitized(payload["model"], maximumLength: Self.maximumModelLength),
                workingDirectory: sanitized(payload["cwd"], maximumLength: Self.maximumPathLength)
            )
        )
    }

    private func parseSessionMetadata(_ root: [String: Any]) -> CodexParsedLine {
        guard let payload = root["payload"] as? [String: Any] else {
            return .malformed("session metadata is missing payload")
        }

        return .sessionMetadata(
            SessionMetadata(
                id: sanitized(payload["id"], maximumLength: Self.maximumIdentifierLength),
                model: sanitized(payload["model"], maximumLength: Self.maximumModelLength),
                workingDirectory: sanitized(payload["cwd"], maximumLength: Self.maximumPathLength),
                forkedFromID: sanitized(payload["forked_from_id"], maximumLength: Self.maximumIdentifierLength),
                parentThreadID: sanitized(payload["parent_thread_id"], maximumLength: Self.maximumIdentifierLength),
                subagentHistoryStartOrdinal: integer(payload["subagent_history_start_ordinal"]),
                occurredAt: (root["timestamp"] as? String).flatMap(parseTimestamp)
            )
        )
    }

    private func parseEventMessage(_ root: [String: Any]) -> CodexParsedLine {
        guard let payload = root["payload"] as? [String: Any],
              let eventType = payload["type"] as? String else {
            return .ignored
        }

        if eventType == "task_started" {
            guard let timestamp = root["timestamp"] as? String,
                  let occurredAt = parseTimestamp(timestamp) else {
                return .malformed("task start is missing a valid timestamp")
            }
            return .taskStarted(
                TaskStartedMetadata(
                    occurredAt: occurredAt,
                    startedAt: timeInterval(payload["started_at"]).map(Date.init(timeIntervalSince1970:)),
                    ordinal: integer(root["ordinal"])
                )
            )
        }

        guard eventType == "token_count" else { return .ignored }

        guard
            let timestamp = root["timestamp"] as? String,
            let occurredAt = parseTimestamp(timestamp)
        else {
            return .malformed("token event is missing a valid timestamp")
        }

        guard let info = payload["info"] as? [String: Any] else {
            return .malformed("token event is missing usage info")
        }

        let lastUsage = parseUsage(info["last_token_usage"])
        let cumulativeUsage = parseUsage(info["total_token_usage"])
        guard lastUsage != nil || cumulativeUsage != nil else {
            return .malformed("token event has no supported usage object")
        }

        return .token(
            CodexTokenObservation(
                occurredAt: occurredAt,
                ordinal: integer(root["ordinal"]),
                lastUsage: lastUsage,
                cumulativeUsage: cumulativeUsage
            )
        )
    }

    private func parseUsage(_ value: Any?) -> TokenUsage? {
        guard let dictionary = value as? [String: Any] else { return nil }
        guard let input = integer(dictionary["input_tokens"]),
              let cached = integer(dictionary["cached_input_tokens"]),
              let output = integer(dictionary["output_tokens"])
        else { return nil }

        let cacheWrite: Int64?
        if dictionary.keys.contains("cache_write_input_tokens") {
            guard let parsed = integer(dictionary["cache_write_input_tokens"]) else { return nil }
            cacheWrite = parsed
        } else {
            cacheWrite = nil
        }

        let usage = TokenUsage(
            inputTokens: input,
            cachedInputTokens: cached,
            cacheWriteInputTokens: cacheWrite,
            outputTokens: output
        )
        return usage.isValid ? usage : nil
    }

    private func integer(_ value: Any?) -> Int64? {
        switch value {
        case let number as NSNumber:
            guard CFGetTypeID(number) != CFBooleanGetTypeID() else { return nil }
            guard !CFNumberIsFloatType(number) else { return nil }
            let result = number.int64Value
            return (0...Self.maximumTokenComponent).contains(result) ? result : nil
        case let string as String:
            guard string.count <= 16 else { return nil }
            return Int64(string).flatMap {
                (0...Self.maximumTokenComponent).contains($0) ? $0 : nil
            }
        default:
            return nil
        }
    }

    private func timeInterval(_ value: Any?) -> TimeInterval? {
        guard let number = value as? NSNumber,
              CFGetTypeID(number) != CFBooleanGetTypeID() else { return nil }
        let seconds = number.doubleValue
        guard seconds.isFinite, seconds >= 0 else { return nil }
        return seconds
    }

    private func sanitized(_ value: Any?, maximumLength: Int) -> String? {
        guard let value = value as? String,
              !value.isEmpty,
              value.count <= maximumLength,
              !value.contains("\0")
        else { return nil }
        return value
    }

    private func parseTimestamp(_ value: String) -> Date? {
        let fractional = Date.ISO8601FormatStyle(includingFractionalSeconds: true)
        if let date = try? fractional.parse(value) {
            return date
        }
        return try? Date.ISO8601FormatStyle().parse(value)
    }
}

File: Windows/src/CodeRim.Core/Parsing/CodexJsonlParser.cs
using System.Globalization;
using System.Text.Json;
using CodeRim.Core.Domain;

namespace CodeRim.Core.Parsing;

public enum ParsedLineKind
{
    SessionMetadata,
    TaskStarted,
    Token,
    Ignored,
    Malformed
}

public sealed record ParsedLine(
    ParsedLineKind Kind,
    SessionMetadata? SessionMetadata = null,
    TaskStartedMetadata? TaskStarted = null,
    TokenObservation? Token = null,
    string? Diagnostic = null);

public static class CodexJsonlParser
{
    public const int MaximumLineBytes = 1_048_576;
    public const long MaximumTokenComponent = 1_000_000_000_000;
    private const int MaximumIdentifierLength = 256;

    public static ParsedLine Parse(ReadOnlyMemory<byte> line)
    {
        if (line.IsEmpty)
        {
            return new ParsedLine(ParsedLineKind.Ignored);
        }

        if (line.Length > MaximumLineBytes)
        {
            return Malformed("line exceeds the 1 MiB safety limit");
        }

        try
        {
            using var document = JsonDocument.Parse(line);
            var root = document.RootElement;
            if (root.ValueKind != JsonValueKind.Object
                || !root.TryGetProperty("type", out var typeElement)
                || typeElement.ValueKind != JsonValueKind.String)
            {
                return Malformed("missing event type");
            }

            return typeElement.GetString() switch
            {
                "session_meta" => ParseSessionMetadata(root),
                "event_msg" => ParseEventMessage(root),
                "turn_context" => new ParsedLine(ParsedLineKind.Ignored),
                _ => new ParsedLine(ParsedLineKind.Ignored)
            };
        }
        catch (JsonException)
        {
            return Malformed("invalid JSON");
        }
    }

    private static ParsedLine ParseSessionMetadata(JsonElement root)
    {
        if (!root.TryGetProperty("payload", out var payload)
            || payload.ValueKind != JsonValueKind.Object)
        {
            return Malformed("session metadata is missing payload");
        }

        return new ParsedLine(
            ParsedLineKind.SessionMetadata,
            SessionMetadata: new SessionMetadata(
                SanitizedString(payload, "id", MaximumIdentifierLength),
                SanitizedString(payload, "forked_from_id", MaximumIdentifierLength),
                SanitizedString(payload, "parent_thread_id", MaximumIdentifierLength),
                TryInteger(payload, "subagent_history_start_ordinal"),
                TryTimestamp(root, "timestamp")));
    }

    private static ParsedLine ParseEventMessage(JsonElement root)
    {
        if (!root.TryGetProperty("payload", out var payload)
            || payload.ValueKind != JsonValueKind.Object
            || !payload.TryGetProperty("type", out var eventType)
            || eventType.ValueKind != JsonValueKind.String)
        {
            return new ParsedLine(ParsedLineKind.Ignored);
        }

        if (eventType.GetString() == "task_started")
        {
            var occurredAt = TryTimestamp(root, "timestamp");
            if (occurredAt is null)
            {
                return Malformed("task start is missing a valid timestamp");
            }

            return new ParsedLine(
                ParsedLineKind.TaskStarted,
                TaskStarted: new TaskStartedMetadata(
                    occurredAt.Value,
                    TryUnixTimestamp(payload, "started_at"),
                    TryInteger(root, "ordinal")));
        }

        if (eventType.GetString() != "token_count")
        {
            return new ParsedLine(ParsedLineKind.Ignored);
        }

        var timestamp = TryTimestamp(root, "timestamp");
        if (timestamp is null)
        {
            return Malformed("token event is missing a valid timestamp");
        }

        if (!payload.TryGetProperty("info", out var info)
            || info.ValueKind != JsonValueKind.Object)
        {
            return Malformed("token event is missing usage info");
        }

        var last = ParseUsage(info, "last_token_usage");
        var cumulative = ParseUsage(info, "total_token_usage");
        if (last is null && cumulative is null)
        {
            return Malformed("token event has no supported usage object");
        }

        return new ParsedLine(
            ParsedLineKind.Token,
            Token: new TokenObservation(
                timestamp.Value,
                TryInteger(root, "ordinal"),
                last,
                cumulative));
    }

    private static TokenUsage? ParseUsage(JsonElement parent, string propertyName)
    {
        if (!parent.TryGetProperty(propertyName, out var usage)
            || usage.ValueKind != JsonValueKind.Object
            || !TryInteger(usage, "input_tokens", out var input)
            || !TryInteger(usage, "cached_input_tokens", out var cached)
            || !TryInteger(usage, "output_tokens", out var output))
        {
            return null;
        }

        var result = new TokenUsage(input, cached, output, TryInteger(usage, "cache_write_input_tokens"));
        return result.IsValid ? result : null;
    }

    private static long? TryInteger(JsonElement parent, string propertyName) =>
        TryInteger(parent, propertyName, out var value) ? value : null;

    private static bool TryInteger(JsonElement parent, string propertyName, out long value)
    {
        value = 0;
        if (!parent.TryGetProperty(propertyName, out var element))
        {
            return false;
        }

        if (element.ValueKind == JsonValueKind.Number)
        {
            return element.TryGetInt64(out value)
                && value >= 0
                && value <= MaximumTokenComponent;
        }

        if (element.ValueKind == JsonValueKind.String)
        {
            var text = element.GetString();
            return text is { Length: <= 16 }
                && long.TryParse(text, NumberStyles.None, CultureInfo.InvariantCulture, out value)
                && value >= 0
                && value <= MaximumTokenComponent;
        }

        return false;
    }

    private static DateTimeOffset? TryTimestamp(JsonElement parent, string propertyName)
    {
        if (!parent.TryGetProperty(propertyName, out var element)
            || element.ValueKind != JsonValueKind.String)
        {
            return null;
        }

        return DateTimeOffset.TryParse(
            element.GetString(),
            CultureInfo.InvariantCulture,
            DateTimeStyles.AssumeUniversal | DateTimeStyles.AdjustToUniversal,
            out var value)
            ? value
            : null;
    }

    private static DateTimeOffset? TryUnixTimestamp(JsonElement parent, string propertyName)
    {
        if (!parent.TryGetProperty(propertyName, out var element)
            || element.ValueKind != JsonValueKind.Number
            || !element.TryGetDouble(out var seconds)
            || !double.IsFinite(seconds)
            || seconds < 0)
        {
            return null;
        }

        try
        {
            return DateTimeOffset.UnixEpoch.AddSeconds(seconds);
        }
        catch (ArgumentOutOfRangeException)
        {
            return null;
        }
    }

    private static string? SanitizedString(JsonElement parent, string propertyName, int maximumLength)
    {
        if (!parent.TryGetProperty(propertyName, out var element)
            || element.ValueKind != JsonValueKind.String)
        {
            return null;
        }

        var value = element.GetString();
        return string.IsNullOrEmpty(value)
            || value.Length > maximumLength
            || value.Contains('\0', StringComparison.Ordinal)
            ? null
            : value;
    }

    private static ParsedLine Malformed(string diagnostic) =>
        new(ParsedLineKind.Malformed, Diagnostic: diagnostic);
}
